core/blockchain.go

```diff
@@ -2808,41 +2804,82 @@ func (bc *BlockChain) UpdateStakingMetaData(tx *staking.StakingTransaction, root
-		return cached.(*big.Int), nil
+		return cached.(*votepower.RewardAccumulation), nil
 	}
 	return rawdb.ReadBlockRewardAccumulator(bc.db, number)
 }
 
 // WriteBlockRewardAccumulator directly writes the BlockRewardAccumulator value
 // Note: this should only be called once during staking launch.
-func (bc *BlockChain) WriteBlockRewardAccumulator(reward *big.Int, number uint64) error {
-	err := rawdb.WriteBlockRewardAccumulator(bc.db, reward, number)
-	if err != nil {
+func (bc *BlockChain) WriteBlockRewardAccumulator(rewarded *economics.Produced) error {
+	if err := rawdb.WriteBlockRewardAccumulator(bc.db, rewarded); err != nil {
 		return err
 	}
-	bc.blockAccumulatorCache.Add(number, reward)
+	bc.blockAccumulatorCache.Add(rewarded.ReadBlockNumber(), rewarded.Read())
 	return nil
 }
 
 //UpdateBlockRewardAccumulator ..
 // Note: this should only be called within the blockchain insertBlock process.
-func (bc *BlockChain) UpdateBlockRewardAccumulator(diff *big.Int, number uint64) error {
-	current, err := bc.ReadBlockRewardAccumulator(number - 1)
+func (bc *BlockChain) UpdateBlockRewardAccumulator(rewarded *economics.Produced) error {
+	current, err := bc.ReadBlockRewardAccumulator(rewarded.ReadBlockNumber() - 1)
+	newWinners := []votepower.VoterReward{}
 	if err != nil {
 		// one-off fix for pangaea, return after pangaea enter staking.
-		current = big.NewInt(0)
-		bc.WriteBlockRewardAccumulator(current, number)
+		bc.WriteBlockRewardAccumulator(economics.NewNoReward(rewarded.ReadBlockNumber()))
+	}
+	rewardsSoFar := current.ValidatorReward
+	soFarCount := len(rewardsSoFar)
+
+	// Sort by address, then can do binary search
+	sort.SliceStable(rewardsSoFar, func(i, j int) bool {
+		return bytes.Compare(
+			rewardsSoFar[i].Validator.Bytes(),
+			rewardsSoFar[j].Validator.Bytes(),
+		) == -1
+	})
+
+	newlyRewarded, existingTotal := rewarded.ReadRewarded(), rewarded.ReadTotalPayout()
+	existingTotal.Add(existingTotal, rewarded.ReadTotalPayout())
+
+	for i := range newlyRewarded {
+		lookingFor := newlyRewarded[i].Validator.Bytes()
+		if k :=
+			sort.Search(soFarCount, func(j int) bool {
+				return bytes.Compare(rewardsSoFar[j].Validator.Bytes(), lookingFor) >= 0
+			}); k < soFarCount &&
+			bytes.Compare(rewardsSoFar[k].Validator.Bytes(), lookingFor) == 0 {
+			// found them, now update
+			for shard := range rewardsSoFar[k].ByShards {
+				for rewardedShard := range newlyRewarded[i].ByShards {
+					if s1, s2 :=
+						rewardsSoFar[k].ByShards[shard],
+						newlyRewarded[i].ByShards[rewardedShard]; s1.ShardID == s2.ShardID {
+						s1.EarnedReward.Add(s1.EarnedReward, s2.EarnedReward)
+					}
+				}
+			}
+		} else {
+			newWinners = append(newWinners, rewardsSoFar[i])
+		}
 	}
-	return bc.WriteBlockRewardAccumulator(new(big.Int).Add(current, diff), number)
+
+	return bc.WriteBlockRewardAccumulator(economics.NewProduced(
+		rewarded.ReadBlockNumber(),
+		append(rewardsSoFar, newWinners...),
```

Comment:
This should be an accumulator for each validator, right? Not just recording rewards for each block. it will blow up the disk.